client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java

```diff
@@ -526,9 +609,29 @@ public boolean isValid(int timeout)
             throws SQLException
     {
         if (timeout < 0) {
-            throw new SQLException("Timeout is negative");
+            throw new SQLException("Timeout is negative.");
+        }
+
+        if (isClosed()) {
+            return false;
+        }
+
+        if (!validateConnection) {
+            return true;
+        }
+
+        try {
+            return validateCredentials(timeout);
+        }
+        catch (UnsupportedOperationException e) {
+            // Backward compatible with older Trino server
+            logger.log(Level.FINE, "Remote server does not support validating connection.", e);
```

Comment:
validateCredentials should not throw

Reply:
In order to return correct SQLState, this method need to return true/false or raise exception. `true` - both connectivity and credentials are valid. `false` - connectivity is valid but credentials are not. Exception - connectivity error. I don't have to raise `UnsupportedOperationException` when server does not support `HEAD`. I can raise `IOException` as well if it is better.